api/envoy/config/core/v3/health_check.proto

```diff
@@ -414,4 +414,12 @@ message HealthCheck {
   // the cluster's :ref:`transport socket <envoy_v3_api_field_config.cluster.v3.Cluster.transport_socket>`
   // will be used for health check socket configuration.
   google.protobuf.Struct transport_socket_match_criteria = 23;
+
+  // When the health status of a host is healthy, if this field is enabled, when envoy wants to send active health check packet to target host, it will first check whether
+  // there is already active business traffic hit the target host during the interval. If yes, no need to send health check packet.
+  // For http, if the business traffic response is 2xx, it will be considered as active traffic
+  // For tcp,  if there is any business traffic, it will be considered as active traffic
+  // For grpc, if grpc_status is non-zero, it will be considered as active traffic
+  // The default value is false.
```

Comment:
```suggestion
  // When the health status of a host is healthy, if this field is enabled, when envoy wants to send an active health check packet to a target host, it will first check whether
  // there was *successful* non-health check traffic that targeted the host during the previous interval. If there was successful non-health check traffic, Envoy will not send a health check packet.
  // For HTTP, if the traffic response is 2xx, it will be considered as successful traffic.
  // For TCP,  if any TCP connections successfully connect, it will be considered as successful traffic.
  // For gRPC, if *grpc-status* is zero, it will be considered as successful traffic
  // The default value is false.
```

You indicated gRPC status as non-zero when I think you meant zero, right?

Reply:
**Filter::handleNon5xxResponseHeaders(absl::optional<Grpc::Status::GrpcStatus> grpc_status,
                                         UpstreamRequest& upstream_request, bool end_stream,
                                         uint64_t grpc_to_http_status) {
  // We need to defer gRPC success until after we have processed grpc-status in
  // the trailers.
  if (grpc_request_) {
    if (end_stream) {
      if (grpc_status && !Http::CodeUtility::is5xx(grpc_to_http_status)) {
        upstream_request.upstreamHost()->stats().rq_success_.inc();
      } else {
        upstream_request.upstreamHost()->stats().rq_error_.inc();
      }**
      From this function, I think when grpc-status is non-zero, it will be consider as successful traffic.

Reply:
I don't think that's what that function does exactly, or it's not implemented correctly. A non-zero grpc-status is an error. Can you double check?

Reply:
From current code, if grpc-status is zero, the request error counter will increase. It seems that it is not correct. 
Just as you said non-zero grpc-status was an error (https://grpc.github.io/grpc/core/md_doc_statuscodes.html). I am really confused with the current logic.

Reply:
In my opinion, the grpc_status checking should be removed.

void Filter::handleNon5xxResponseHeaders(absl::optional<Grpc::Status::GrpcStatus> grpc_status,
                                         UpstreamRequest& upstream_request, bool end_stream,
                                         uint64_t grpc_to_http_status) {
  // We need to defer gRPC success until after we have processed grpc-status in
  // the trailers.
  if (grpc_request_) {
    if (end_stream) {
      if ( !Http::CodeUtility::is5xx(grpc_to_http_status)) {
        upstream_request.upstreamHost()->stats().rq_success_.inc();
        upstream_request.upstreamHost()->setLastSuccessfulTrafficTime(
            envoy::data::core::v3::HealthCheckerType::GRPC,
            callbacks_->dispatcher().timeSource().monotonicTime());
      } else {
        upstream_request.upstreamHost()->stats().rq_error_.inc();
      }
    } else {
      upstream_request.grpcRqSuccessDeferred(true);
    }
  } else {
    upstream_request.upstreamHost()->stats().rq_success_.inc();
  }